<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU34_Predicción de demanda de servicios</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 01. Simulación de demanda de servicios

* Simulación de demanda de servicios durante un periodo de tiempo por sección censal
* Se toman unas fechas de inicio y fin y un conjunto de servicios
* Se expanden a secciones censales y se simulan demandas de servicios







<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Encoding

Con la siguiente expresión se evitan problemas con el encoding al ejecutar el notebook. Es posible que deba ser eliminada o adaptada a la máquina en la que se ejecute el código.

In [1]:
Sys.setlocale(category = "LC_ALL", locale = "es_ES.UTF-8")

[1] "es_ES.UTF-8/es_ES.UTF-8/es_ES.UTF-8/C/es_ES.UTF-8/C"

### Packages to use

_ELIMINAR O AÑADIR LO QUE TOQUE. COPIAR VERSIONES AL FINAL Y QUITAR CÓDIGO DE VERSIONES_

- {tcltk} para selección interactiva de archivos locales
- {sf} para trabajar con georeferenciación
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos
- {stringr} para manipulación de cadenas de caracteres
- {tidyr} para organización de datos


In [2]:
library(sf)
library(readr)
library(dplyr)
library(lubridate)
library(elevatr)




Linking to GEOS 3.10.2, GDAL 3.4.2, PROJ 8.2.1; sf_use_s2() is TRUE


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



Attaching package: ‘lubridate’


The following objects are masked from ‘package:base’:

    date, intersect, setdiff, union




### Paths

In [3]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

1. Geometrías de las secciones censales

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [4]:
# file_data_01 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [5]:
iFile_01 <- "CU_34_05_01_secciones_geo.json"
file_data_01 <- paste0(iPath, iFile_01)

if(file.exists(file_data_01)){
    cat("Se leerán datos del archivo: ", file_data_01)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leerán datos del archivo:  Data/Input/CU_34_05_01_secciones_geo.json

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [6]:
data_01 <- st_read(file_data_01)


Reading layer `CU_34_05_01_secciones_geo' from data source 
  `/Users/emilio.lcano/academico/gh_repos/__transferencia/citizenlab/CitizenLab-Research-and-Development/casos_urjc/notebooks/II_data_processing/34_servicios/Data/Input/CU_34_05_01_secciones_geo.json' 
  using driver `GeoJSON'
Simple feature collection with 4432 features and 16 fields
Geometry type: MULTIPOLYGON
Dimension:     XY
Bounding box:  xmin: -4.579006 ymin: 39.8848 xmax: -3.052983 ymax: 41.16584
Geodetic CRS:  WGS 84


Estructura de  los datos:

In [7]:
data_01 |> glimpse()

Rows: 4,432
Columns: 17
$ CUSEC    <chr> "2800101001", "2800201001", "2800201002", "2800301001", "2800…
$ CUMUN    <chr> "28001", "28002", "28002", "28003", "28004", "28004", "28004"…
$ CSEC     <chr> "001", "001", "002", "001", "001", "002", "003", "004", "001"…
$ CDIS     <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01", "01", "…
$ CMUN     <chr> "001", "002", "002", "003", "004", "004", "004", "004", "005"…
$ CPRO     <chr> "28", "28", "28", "28", "28", "28", "28", "28", "28", "28", "…
$ CCA      <chr> "13", "13", "13", "13", "13", "13", "13", "13", "13", "13", "…
$ CUDIS    <chr> "2800101", "2800201", "2800201", "2800301", "2800401", "28004…
$ CLAU2    <chr> "28001", "28002", "28002", "28003", "28004", "28004", "28004"…
$ NPRO     <chr> "Madrid", "Madrid", "Madrid", "Madrid", "Madrid", "Madrid", "…
$ NCA      <chr> "Comunidad de Madrid", "Comunidad de Madrid", "Comunidad de M…
$ CNUT0    <chr> "ES", "ES", "ES", "ES", "ES", "ES", "ES", "ES", "ES", "ES", "…
$ CNUT1    <chr>

Muestra de datos:

In [8]:
data_01 |> tibble() |> slice_head(n = 5)

CUSEC      CUMUN CSEC CDIS CMUN CPRO CCA CUDIS   CLAU2 NPRO  
1 2800101001 28001 001  01   001  28   13  2800101 28001 Madrid
2 2800201001 28002 001  01   002  28   13  2800201 28002 Madrid
3 2800201002 28002 002  01   002  28   13  2800201 28002 Madrid
4 2800301001 28003 001  01   003  28   13  2800301 28003 Madrid
5 2800401001 28004 001  01   004  28   13  2800401 28004 Madrid
  NCA                 CNUT0 CNUT1 CNUT2 CNUT3 NMUN             
1 Comunidad de Madrid ES    3     0     0     Acebeda, La      
2 Comunidad de Madrid ES    3     0     0     Ajalvir          
3 Comunidad de Madrid ES    3     0     0     Ajalvir          
4 Comunidad de Madrid ES    3     0     0     Alameda del Valle
5 Comunidad de Madrid ES    3     0     0     Álamo, El        
  geometry                      
1 MULTIPOLYGON (((-3.64316 41...
2 MULTIPOLYGON (((-3.469465 4...
3 MULTIPOLYGON (((-3.503032 4...
4 MULTIPOLYGON (((-3.808662 4...
5 MULTIPOLYGON (((-3.979038 4...

2. Indicadores de las secciones censales

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [9]:
# file_data_02 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [10]:
iFile_02 <- "CU_04_05_03_01_indicadores_secciones.csv"
file_data_02 <- paste0(iPath, iFile_02)

if(file.exists(file_data_02)){
    cat("Se leerán datos del archivo: ", file_data_02)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leerán datos del archivo:  Data/Input/CU_04_05_03_01_indicadores_secciones.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [11]:
data_02 <- read_csv(file_data_02)


Rows: 4417 Columns: 20
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): CMUN, dist, secc
dbl (17): ccaa, CPRO, t1_1, t2_1, t2_2, t3_1, t4_1, t4_2, t4_3, t5_1, t6_1, ...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [12]:
data_02|> glimpse()

Rows: 4,417
Columns: 20
$ ccaa  <dbl> 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, …
$ CPRO  <dbl> 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, …
$ CMUN  <chr> "001", "002", "002", "003", "004", "004", "004", "004", "005", "…
$ dist  <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01", "01", "01"…
$ secc  <chr> "001", "001", "002", "001", "001", "002", "003", "004", "001", "…
$ t1_1  <dbl> 55, 2358, 2435, 248, 2886, 3376, 2161, 1523, 1648, 1015, 1728, 1…
$ t2_1  <dbl> NA, 0.4724, 0.4830, 0.3952, 0.5135, 0.4793, 0.5016, 0.5279, 0.54…
$ t2_2  <dbl> NA, 0.5276, 0.5170, 0.6048, 0.4865, 0.5207, 0.4984, 0.4721, 0.45…
$ t3_1  <dbl> NA, 40.5161, 38.3339, 47.4597, 44.4099, 40.6810, 38.0088, 42.537…
$ t4_1  <dbl> NA, 0.1425, 0.1959, 0.1371, 0.1611, 0.1739, 0.2230, 0.1753, 0.11…
$ t4_2  <dbl> NA, 0.7443, 0.7002, 0.6573, 0.6067, 0.6727, 0.6395, 0.6448, 0.69…
$ t4_3  <dbl> NA, 0.1132, 0.1039, 0.2056, 0.2322, 0.1534, 0.1374, 0.1799, 0.18…
$ t5_1  <dbl> NA

Muestra de datos:

In [13]:
data_02 |> tibble() |> slice_head(n = 5)

ccaa CPRO CMUN dist secc t1_1 t2_1   t2_2   t3_1    t4_1   t4_2   t4_3  
1 13   28   001  01   001    55     NA     NA      NA     NA     NA     NA
2 13   28   002  01   001  2358 0.4724 0.5276 40.5161 0.1425 0.7443 0.1132
3 13   28   002  01   002  2435 0.4830 0.5170 38.3339 0.1959 0.7002 0.1039
4 13   28   003  01   001   248 0.3952 0.6048 47.4597 0.1371 0.6573 0.2056
5 13   28   004  01   001  2886 0.5135 0.4865 44.4099 0.1611 0.6067 0.2322
  t5_1   t6_1   t7_1   t8_1   t9_1   t10_1  t11_1  t12_1 
1     NA     NA     NA     NA     NA     NA     NA     NA
2 0.1569 0.1968 0.4016 0.3971 0.4377 0.0912 0.6063 0.6672
3 0.1544 0.1951 0.4244 0.4224 0.4438 0.1226 0.5955 0.6788
4 0.1129 0.1411 0.1542 0.1449 0.5280 0.0932 0.5000 0.5514
5 0.1639 0.2128 0.1470 0.1409 0.2602 0.1814 0.4213 0.5147

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Secciones censales de la Comunidad de Madrid
* Indicadores del INE por sección censal

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

Si no aplica: Estos datos no requieren tareas de este tipo.

#### Data transformation

* Asignar nombre de sección
* Calcular área
* Obtener elevación
* Eliminar geometrías

In [14]:
tdata_01 <- data_01 |>
    mutate(NSEC = paste0(NMUN, " - ", CDIS, ".", CSEC),
        area = st_area(data_01)) |>
    st_centroid() |>
    get_elev_point(src = "aws")  |>
    select(-elev_units) |>
    st_drop_geometry() 


Warning message in st_centroid.sf(mutate(data_01, NSEC = paste0(NMUN, " - ", CDIS, :
“st_centroid assumes attributes are constant over geometries of x”
Mosaicing & Projecting

Note: Elevation units are in meters



In [15]:
tdata_01 |> glimpse()

Rows: 4,432
Columns: 19
$ CUSEC     <chr> "2800101001", "2800201001", "2800201002", "2800301001", "280…
$ CUMUN     <chr> "28001", "28002", "28002", "28003", "28004", "28004", "28004…
$ CSEC      <chr> "001", "001", "002", "001", "001", "002", "003", "004", "001…
$ CDIS      <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01", "01", …
$ CMUN      <chr> "001", "002", "002", "003", "004", "004", "004", "004", "005…
$ CPRO      <chr> "28", "28", "28", "28", "28", "28", "28", "28", "28", "28", …
$ CCA       <chr> "13", "13", "13", "13", "13", "13", "13", "13", "13", "13", …
$ CUDIS     <chr> "2800101", "2800201", "2800201", "2800301", "2800401", "2800…
$ CLAU2     <chr> "28001", "28002", "28002", "28003", "28004", "28004", "28004…
$ NPRO      <chr> "Madrid", "Madrid", "Madrid", "Madrid", "Madrid", "Madrid", …
$ NCA       <chr> "Comunidad de Madrid", "Comunidad de Madrid", "Comunidad de …
$ CNUT0     <chr> "ES", "ES", "ES", "ES", "ES", "ES", "ES", "ES", "ES", "ES", …
$ CNUT1     <chr

* Unir área y elevación a la tabla de indicadores, y eliminar geometría

In [16]:
tdata_02 <- data_02 |>
    full_join(tdata_01 |> select(NSEC, CMUN, CSEC, CDIS, area, elevation),
        by = c("CMUN" = "CMUN", "dist" = "CDIS", "secc" = "CSEC")) |>
    mutate(densidad_hab_km2 = as.numeric(t1_1/(area*1e-6))) |>
    st_drop_geometry() 

In [17]:
tdata_02 |> glimpse()

Rows: 4,440
Columns: 24
$ ccaa             <dbl> 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 1…
$ CPRO             <dbl> 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 2…
$ CMUN             <chr> "001", "002", "002", "003", "004", "004", "004", "004…
$ dist             <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01",…
$ secc             <chr> "001", "001", "002", "001", "001", "002", "003", "004…
$ t1_1             <dbl> 55, 2358, 2435, 248, 2886, 3376, 2161, 1523, 1648, 10…
$ t2_1             <dbl> NA, 0.4724, 0.4830, 0.3952, 0.5135, 0.4793, 0.5016, 0…
$ t2_2             <dbl> NA, 0.5276, 0.5170, 0.6048, 0.4865, 0.5207, 0.4984, 0…
$ t3_1             <dbl> NA, 40.5161, 38.3339, 47.4597, 44.4099, 40.6810, 38.0…
$ t4_1             <dbl> NA, 0.1425, 0.1959, 0.1371, 0.1611, 0.1739, 0.2230, 0…
$ t4_2             <dbl> NA, 0.7443, 0.7002, 0.6573, 0.6067, 0.6727, 0.6395, 0…
$ t4_3             <dbl> NA, 0.1132, 0.1039, 0.2056, 0.2322, 0.1534, 0.1374, 0…
$ t5_1          

## Fake Data Generation

* Se crea una función para simular el número de servicios de cada uno de los dos tipos propuestos: 
    - Taxi y delivery
    - Dependen de si hay evento importante (fútbol) del número de titulados y de la densidad de población
* Después se simula a lo largo del periodo parametrizado

* Parámetros:

In [18]:
fechas <- seq(from = as.Date("2022-01-01"), 
                             to = as.Date("2022-01-31"),
              by = 1)

servicios <- c("Taxi", "Delivery")

* Funciones:

In [19]:
futbol <- Vectorize(
  function(fecha, servicio){
    p <- 0.8
    if(wday(fecha) %in% 6:7){
      p <- p*(1+0.1*(servicio == "Taxi"))
    } 
    sample(0:1, 1, prob = c(p, 1-p))
  })

n_servicios <- Vectorize(
  function(futbol, servicio, titulados, densidad){
    titulados <- ifelse(is.na(titulados), 0, titulados)
    densidad <- ifelse(is.na(densidad), 1e-6, densidad)
    l <- rpois(1, round(densidad*1e-6*100)) + 50*futbol
    if(servicio == "Delivery" & !is.na(titulados)){
      l <- l + round(20*titulados)
    }
    return(l)
  }
)

* Simulación de datos fake demanda de servicio

In [20]:
set.seed(1)
fdata_01 <- expand.grid(NSEC = tdata_01$NSEC, 
                  Fecha = fechas, 
                  Servicio = servicios) |> 
  inner_join(tdata_01 |> 
               select(NSEC, CMUN, CDIS, CSEC, area, elevation)) |> 
  inner_join(tdata_02 |> 
               select(personas = t1_1, 
                      edad = t3_1,
                      mayores = t4_3,
                      titulados = t9_1,
                      CMUN, dist, secc, densidad_hab_km2),
             by = c("CMUN" = "CMUN", 
                    "CDIS" = "dist",
                    "CSEC" = "secc")) |> 
  mutate(Futbol = futbol(Fecha, Servicio)) |> 
  mutate(nservicios = n_servicios(Futbol, Servicio, titulados, densidad_hab_km2)) |>
  select(-NSEC, -area, -elevation, -personas, -edad, -mayores, -titulados, -densidad_hab_km2)


Joining with `by = join_by(NSEC)`


In [21]:
fdata_01 |> glimpse()

Rows: 274,784
Columns: 7
$ Fecha      <date> 2022-01-01, 2022-01-01, 2022-01-01, 2022-01-01, 2022-01-01…
$ Servicio   <fct> Taxi, Taxi, Taxi, Taxi, Taxi, Taxi, Taxi, Taxi, Taxi, Taxi,…
$ CMUN       <chr> "001", "002", "002", "003", "004", "004", "004", "004", "00…
$ CDIS       <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01", "01",…
$ CSEC       <chr> "001", "001", "002", "001", "001", "002", "003", "004", "00…
$ Futbol     <int> 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,…
$ nservicios <dbl> 0, 0, 0, 50, 0, 50, 50, 0, 2, 0, 0, 7, 0, 1, 4, 7, 10, 50, …


In [22]:
fdata_01 |> slice_head(n = 5)

Fecha      Servicio CMUN CDIS CSEC Futbol nservicios
1 2022-01-01 Taxi     001  01   001  0       0        
2 2022-01-01 Taxi     002  01   001  0       0        
3 2022-01-01 Taxi     002  01   002  0       0        
4 2022-01-01 Taxi     003  01   001  1      50        
5 2022-01-01 Taxi     004  01   001  0       0

* Simulación datos fake capacidad de servicios. Asumimos capacidad por municipio, el máximo de todos los días

In [23]:
fdata_02 <- fdata_01 |> 
    group_by(CMUN, Servicio) |>
    summarise(capacidad = max(nservicios), .groups = "drop")

In [24]:
glimpse(fdata_02)

Rows: 358
Columns: 3
$ CMUN      <chr> "001", "001", "002", "002", "003", "003", "004", "004", "005…
$ Servicio  <fct> Taxi, Delivery, Taxi, Delivery, Taxi, Delivery, Taxi, Delive…
$ capacidad <dbl> 50, 50, 50, 59, 50, 61, 50, 56, 65, 70, 66, 75, 64, 70, 50, …


In [25]:
fdata_02 |> slice_head(n = 5)

CMUN Servicio capacidad
1 001  Taxi     50       
2 001  Delivery 50       
3 002  Taxi     50       
4 002  Delivery 59       
5 003  Taxi     50

## Synthetic Data Generation

No aplica

## Open Data

No aplica

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

1. Indicadores INE con identificador sección, área, elevación y densidad

<font color='tomato'> Identificamos los datos a guardar</font>

In [26]:
data_to_save_01 <- tdata_02


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [27]:
caso <- "CU_34"
proceso <- '_05'
tarea <- "_02"
archivo <- "_01"
proper <- "_indicadores_secc_extendido"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [28]:
# file_save_01 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_01 <- paste0(oPath, file_save_01)
# write_csv(data_to_save_01, path_out_01)

# cat('File saved as: ')
# path_out_01

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [29]:
file_save_01 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_01 <- paste0(oPath, file_save_01)
write_csv(data_to_save_01, path_out_01)

cat('File saved as: ')
path_out_01

File saved as: 

[1] "Data/Output/CU_34_05_02_01_indicadores_secc_extendido.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [30]:
path_in_01 <- paste0(iPath, file_save_01)
file.copy(path_out_01, path_in_01, overwrite = TRUE)

[1] TRUE

2. Simulación de demanda de servicios

<font color='tomato'> Identificamos los datos a guardar</font>

In [31]:
data_to_save_02 <- fdata_01


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [32]:
archivo <- "_02"
proper <- "_demanda_sim"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [33]:
# file_save_02 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_02 <- paste0(oPath, file_save_02)
# write_csv(data_to_save_02, path_out_02)

# cat('File saved as: ')
# path_out_02

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [34]:
file_save_02 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_02 <- paste0(oPath, file_save_02)
write_csv(data_to_save_02, path_out_02)

cat('File saved as: ')
path_out_02

File saved as: 

[1] "Data/Output/CU_34_05_02_02_demanda_sim.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [35]:
path_in_02 <- paste0(iPath, file_save_02)
file.copy(path_out_02, path_in_02, overwrite = TRUE)

[1] TRUE

3. Simulación de capacidad de servicios

<font color='tomato'> Identificamos los datos a guardar</font>

In [36]:
data_to_save_03 <- fdata_02


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [37]:
archivo <- "_03"
proper <- "_capacidad_sim"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [38]:
# file_save_03 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_03 <- paste0(oPath, file_save_03)
# write_csv(data_to_save_02, path_out_02)

# cat('File saved as: ')
# path_out_03

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [39]:
file_save_03 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_03 <- paste0(oPath, file_save_03)
write_csv(data_to_save_03, path_out_03)

cat('File saved as: ')
path_out_03

File saved as: 

[1] "Data/Output/CU_34_05_02_03_capacidad_sim.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [40]:
path_in_03 <- paste0(iPath, file_save_03)
file.copy(path_out_03, path_in_03, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

Para que funcione este código se necesita:

* Las rutas de archivos `Data/Input` y `Data/Output` deben existir (relativas a la ruta del _notebook_)
* El paquete tcltk instalado para seleccionar archivos interactivamente. No se necesita en producción.
* Los paquetes sf, readr, dplyr, lubridate, elevatr deben estar instalados.


### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* sf 1.0.9
* readr 2.1.3
* dplyr 1.1.0
* lubridate 1.9.1
* elevatr 0.4.2

#### Objeto `tdata_02`

- Datos de indicadores ampliados
- Hay 4440 filas con información de las siguientes variables:
	* ccaa
	* CPRO
	* CMUN
	* dist
	* secc
	* t1_1
	* t2_1
	* t2_2
	* t3_1
	* t4_1
	* t4_2
	* t4_3
	* t5_1
	* t6_1
	* t7_1
	* t8_1
	* t9_1
	* t10_1
	* t11_1
	* t12_1
	* NSEC
	* area
	* elevation
	* densidad_hab_km2

#### Objeto `fdata_01`

- Datos de demanda de servicios
- Hay 274784 filas con información de las siguientes variables:
	* Fecha
	* Servicio
	* CMUN
	* CDIS
	* CSEC
	* area
	* elevation
	* personas
	* edad
	* mayores
	* titulados
	* densidad_hab_km2
	* Futbol
	* nservicios

#### Objeto `fdata_02`

- Datos de capacidad de servicios
- Hay 358 filas con información de las siguientes variables:
	* CMUN
	* Servicio
	* capacidad

#### Observaciones generales sobre los datos

* No aplica

### Consideraciones para despliegue en piloto

* Los datos son simulados

### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para obtener datos reales

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se ha asignado una etiqueta a cada sección censal
- Se ha calculado la densidad de población de cada sección
- Se ha calculado la elevación de las secciones censales
- Se han simulado datos de servicios

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben añadir los datos meteorológicos

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [41]:
# incluir código